cmake/modules/zephyr_default.cmake

```diff
@@ -32,41 +32,42 @@ find_package(ZephyrBuildConfiguration
 )
 
 # Prepare user cache
-include(python)
-include(user_cache)
+list(APPEND sub_components python)
+list(APPEND sub_components user_cache)
 
 # Load Zephyr extensions
-include(extensions)
-include(git)
-include(version)  # depends on hex.cmake
+list(APPEND sub_components extensions)
+list(APPEND sub_components git)
+list(APPEND sub_components version)
 
 #
 # Find tools
 #
 
-include(west)
-include(ccache)
+list(APPEND sub_components west)
+list(APPEND sub_components ccache)
 
 # Load default root settings
-include(root)
+list(APPEND sub_components root)
 
 #
 # Find Zephyr modules.
 # Those may contain additional DTS, BOARD, SOC, ARCH ROOTs.
 # Also create the Kconfig binary dir for generated Kconf files.
 #
-include(zephyr_module)
+list(APPEND sub_components zephyr_module)
 
-include(boards)
-include(shields)
-include(arch)
-include(configuration_files)
+list(APPEND sub_components boards)
+list(APPEND sub_components shields)
+list(APPEND sub_components arch)
+list(APPEND sub_components configuration_files)
 
-include(verify-toolchain)
-include(host-tools)
+list(APPEND sub_components verify-toolchain)
+list(APPEND sub_components host-tools)
 
 # Include board specific device-tree flags before parsing.
-include(${BOARD_DIR}/pre_dt_board.cmake OPTIONAL)
+set(pre_dt_board "\${BOARD_DIR}/pre_dt_board.cmake" OPTIONAL)
+list(APPEND sub_components "\${pre_dt_board}")
```

Comment:
I don't follow how this works with the loop below. How are you handling the OPTIONAL part?

Reply:
This is handled here:
https://github.com/zephyrproject-rtos/zephyr/blob/8f2599cdb1284dba80532a2c7cf580ffb884b796/cmake/modules/zephyr_default.cmake#L117-L118

First time, the `${pre_dt_board}` from here: https://github.com/zephyrproject-rtos/zephyr/blob/8f2599cdb1284dba80532a2c7cf580ffb884b796/cmake/modules/zephyr_default.cmake#L70
gets expanded, second time the `\${BOARD_DIR}` from here https://github.com/zephyrproject-rtos/zephyr/blob/8f2599cdb1284dba80532a2c7cf580ffb884b796/cmake/modules/zephyr_default.cmake#L69
gets expanded.

Which results in the following code being executed. (CMake trace output)
```
.../zephyr/cmake/modules/zephyr_default.cmake(117):  string(CONFIGURE ${pre_dt_board} module )
.../zephyr/cmake/modules/zephyr_default.cmake(118):  string(CONFIGURE ${BOARD_DIR}/pre_dt_board.cmake;OPTIONAL module )
.../zephyr/cmake/modules/zephyr_default.cmake(119):  include(/projects/github/ncs/zephyr/boards/arm/nrf52840dk_nrf52840/pre_dt_board.cmake;OPTIONAL )
```

so as you see, the `OPTIONAL` is expanded as expected and passed to the `include()` call as argument.